libraries/networking/src/webrtc/WebRTCDataChannels.cpp

```diff
@@ -0,0 +1,492 @@
+    SdpParseError sdpParseError;
+    auto iceCandidate = CreateIceCandidate(
+        data.value("sdpMid").toString().toStdString(),
+        data.value("sdpMLineIndex").toInt(),
+        data.value("candidate").toString().toStdString(),
+        &sdpParseError);
+    if (!iceCandidate) {
+        qCWarning(networking_webrtc) << "Error adding WebRTC ICE candidate:"
+            << QString::fromStdString(sdpParseError.description);
+        return;
+    }
+
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "6. Add ICE candidate";
+#endif
+    _peerConnection->AddIceCandidate(iceCandidate);
+}
+
+void WDCConnection::sendIceCandidate(const IceCandidateInterface* candidate) {
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "WDCConnection::sendIceCandidate()";
+#endif
+
+    std::string candidateString;
+    candidate->ToString(&candidateString);
+    QJsonObject jsonCandidate;
+    jsonCandidate.insert("candidate", QString::fromStdString(candidateString));
+    jsonCandidate.insert("sdpMid", QString::fromStdString(candidate->sdp_mid()));
+    jsonCandidate.insert("sdpMLineIndex", candidate->sdp_mline_index());
+
+    QJsonObject jsonWebRTCData;
+    jsonWebRTCData.insert("candidate", jsonCandidate);
+
+    QJsonObject jsonObject;
+    jsonObject.insert("from", QString(_parent->getNodeType()));
+    jsonObject.insert("to", _webSocketID);
+    jsonObject.insert("data", jsonWebRTCData);
+    QJsonDocument jsonDocument = QJsonDocument(jsonObject);
+
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "7. Send ICE candidate to the remote peer";
+#endif
+    _parent->sendSignalingMessage(jsonObject);
+}
+
+void WDCConnection::onDataChannelOpened(rtc::scoped_refptr<DataChannelInterface> dataChannel) {
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "WDCConnection::onDataChannelOpened() :"
+        << dataChannel->id()
+        << QString::fromStdString(dataChannel->label())
+        << QString::fromStdString(dataChannel->protocol())
+        << dataChannel->negotiated()
+        << dataChannel->maxRetransmitTime()
+        << dataChannel->maxRetransmits()
+        << dataChannel->maxPacketLifeTime().value_or(-1)
+        << dataChannel->maxRetransmitsOpt().value_or(-1);
+#endif
+
+    _dataChannel = dataChannel;
+    _dataChannelID = _parent->getNewDataChannelID();  // Not dataChannel->id() because it's only unique per peer connection.
+    _dataChannel->RegisterObserver(_dataChannelObserver.get());
+
+    _parent->onDataChannelOpened(this, _dataChannelID);
+}
+
+void WDCConnection::onDataChannelStateChanged() {
+    auto state = _dataChannel->state();
+#ifdef WEBRTC_DEBUG
+    qCDebug(networking_webrtc) << "WDCConnection::dataChannelStateChanged() :" << (int)state
+        << DataChannelInterface::DataStateString(state);
+#endif
+    if (state == DataChannelInterface::kClosed) {
+        _dataChannel->Close();
+        _dataChannel = nullptr;
+        // WEBRTC FIXME: The following line causes the _peerConnectionFactory to fail.
```

Comment:
What happens then? Do we have a leak here meanwhile?

Reply:
Fixed